Creating a ROS Index Tutorial

CursedRock17 opened the pull request:
Meant to resolve #2801 by adding a tutorial for using [ROS Index](https://index.ros.org/), that way there's more clarity on beginners finding ROS 2 packages they'll need. I've originally stuck this tutorial in the `How-To-Guides` section, but since it's for beginners it may be more appropriate to put in the `Beginners` section of the Tutorials, perhaps making a new section called tooling. It's meant to be short and sweet - to clarify efficiently - so using pictures will help them understand.

CursedRock17 committed:
Creating a ROS Index Tutorial

Signed-off-by: CursedRock17 <mtglucas1@gmail.com>

CursedRock17 committed:
Fixing a typo

Signed-off-by: CursedRock17 <mtglucas1@gmail.com>

ahcorde reviewed (changes requested):
check linters

clalancette commented:
I have to say that I'm not completely sure we should have a tutorial like this.  This is essentially "how to navigate a web page", and we have to expect some level of technical familiarity from our audience.

I think that if this stuff is unclear, we should fix the presentation on https://index.ros.org to make it more clear.  Having a tutorial here just seems odd.  But that is my two cents; @ros2/team , other thoughts?

CursedRock17 commented:
I don't know if the website is that unclear though, I think the original issuer may be looking for more tutorials on their issues (which I was planning to fix today). The only thing that might be unclear is being able to search for keywords, somebody else has put down that we have the backend for it. The original issuer might not have known that you don't have to do just package names.

kscottz commented on source/How-To-Guides/Using-ROSIndex.rst:
```diff
@@ -0,0 +1,51 @@
+2. `Tutorials` If there are any tutorials available in the documentation they will show up here.
+Those looking to get started with a package should look here as the tutorials will often be up to date and well maintained by those who use the package often.
+3. `Documentation` The original documentation and github repository where this data is found
+4. `Distrobution` ROS 2 has `various distrobutions <>https://docs.ros.org/en/rolling/Releases.html ` which may have differences in design.
```
```suggestion
4. `Distribution` ROS 2 has `various distributions <>https://docs.ros.org/en/rolling/Releases.html ` which may have differences in design.
```

kscottz commented on source/How-To-Guides/Using-ROSIndex.rst:
```diff
@@ -0,0 +1,51 @@
+   :local:
+
+
+This guide introduces the standard way of searching for documentation among ROS 2 packages.
```
```suggestion
This guide introduces the standard way of searching for ROS 2 package-level documentation.
```

kscottz commented:
I think this is a great beginner tutorial. We have a bit of a "librarian" problem where new users don't know about the resources available to them because they are spread out over half a dozen different websites. New ROS users do not understand that ROSIndex exists and that it is effectively PyPI for ROS packages.  Given the current state of package-level docs on docs.ros.org we really need users using ROS Index to access their package-level documentation.  

I think a tutorial on this topic is completely fine. I do think this PR could use improvement. I'll send over some recommendations shortly.

kscottz commented on source/How-To-Guides/Using-ROSIndex.rst:
```diff
@@ -0,0 +1,51 @@
+
+
+This guide introduces the standard way of searching for documentation among ROS 2 packages.
+The best way to find solutions to one's problems will be using ROS Index to crawl through packages in the ROS 2 ecosystem.
```
```suggestion
There are thousands of ROS packages that encompass all aspects of robotics; everything from hardware drivers to high-level planning and navigation. 
Some of these packages exist solely as open source code, while others are distributed as binary packages that can be easily installed on a ROS system using tools like `apt`. 
A complete and searchable list of these ROS binary packages is available at `index.ros.org <http://index.ros.org>`__, which is similar to `Python's Package Index (PyPI) <https://pypi.org/>`__.
A good way to find a solution to many common robotics problems is to use ROS Index to find an appropriate ROS 2 binary package from the ROS ecosystem.
```

kscottz commented on source/How-To-Guides/Using-ROSIndex.rst:
```diff
@@ -0,0 +1,51 @@
+
+This guide introduces the standard way of searching for documentation among ROS 2 packages.
+The best way to find solutions to one's problems will be using ROS Index to crawl through packages in the ROS 2 ecosystem.
+One should find packages that would greatly benefit them and using the tutorials that go along with these packages, to increase one's understanding about the topic.
```
```suggestion
Using ROS Index you can find binary ROS packages, the source code behind them, and their associated documentation  and  tutorials.
```

kscottz commented on source/How-To-Guides/Using-ROSIndex.rst:
```diff
@@ -0,0 +1,51 @@
+- `Open ROS Index <https://index.ros.org/>`
+
+
+Finding a Beneficial Package
```
```suggestion
Searching for ROS Packages
```

kscottz commented on source/How-To-Guides/Using-ROSIndex.rst:
```diff
@@ -0,0 +1,51 @@
+Finding a Beneficial Package
+----------------------------
+
+Locate the search bar in the middle of the screen:
```
```suggestion
Locate the search bar in the middle of the screen on `index.ros.org <https://index.ros.org/>`__:
```

kscottz commented on source/How-To-Guides/Using-ROSIndex.rst:
```diff
@@ -0,0 +1,51 @@
+  :width: 500 px
+  :alt: Search Bar on ROS Index
+
+Then utilize this to find either a specific package or search for the features of a package.
```
```suggestion
You can use this search bar to find specific packages or search for specific package features.
```

kscottz commented on source/How-To-Guides/Using-ROSIndex.rst:
```diff
@@ -0,0 +1,51 @@
+  :alt: Search Bar on ROS Index
+
+Then utilize this to find either a specific package or search for the features of a package.
+ROS Index has the capabilities to search through the available documentation and find keywords that match the given search.
```
```suggestion
ROS Index has the capability to search through the available package documentation and find keywords that match the given search query.
```

kscottz commented on source/How-To-Guides/Using-ROSIndex.rst:
```diff
@@ -0,0 +1,51 @@
+
+Then utilize this to find either a specific package or search for the features of a package.
+ROS Index has the capabilities to search through the available documentation and find keywords that match the given search.
+If interested, one can `Create Documentation <https://docs.ros.org/en/rolling/How-To-Guides/Documenting-a-ROS-2-Package.html>` for their own packages.
```
Let's drop this line for now and focus on finding packages. 
```suggestion
```

kscottz commented on source/How-To-Guides/Using-ROSIndex.rst:
```diff
@@ -0,0 +1,51 @@
+Those looking to get started with a package should look here as the tutorials will often be up to date and well maintained by those who use the package often.
+3. `Documentation` The original documentation and github repository where this data is found
+4. `Distrobution` ROS 2 has `various distrobutions <>https://docs.ros.org/en/rolling/Releases.html ` which may have differences in design.
+One should click the distrobution which most fits their needs.
```
```suggestion
Click on the ROS distribution (version) that matches the ROS version installed on your system. 
```

kscottz commented on source/How-To-Guides/Using-ROSIndex.rst:
```diff
@@ -0,0 +1,51 @@
+One should click the distrobution which most fits their needs.
+5. `Contributing` This software is open source, so it needs the community's help to actively be maintained.
+One should utilize this section if they wish to help solve issues, review pull requests, or are looking to see if an active fix is being deployed for a problem they encountered.
+6. `Maintainers` Those in the community actively responsible for the development of the package.
```
```suggestion
6. `Maintainers` This tab lists the community members who are actively responsible for the development of the package.
```

kscottz commented on source/How-To-Guides/Using-ROSIndex.rst:
```diff
@@ -0,0 +1,51 @@
+1. `Package Description` The general description of the package, this should be similar to what was searched if it wasn't an exact package name.
+2. `Tutorials` If there are any tutorials available in the documentation they will show up here.
+Those looking to get started with a package should look here as the tutorials will often be up to date and well maintained by those who use the package often.
+3. `Documentation` The original documentation and github repository where this data is found
```
```suggestion
3. `Documentation` The original documentation for the ROS package and the repository where source code and documentation can be found.```

kscottz reviewed (changes requested):
I threw in a bunch of comments and cleanup. Let's merge these in and see where we're at. I imagine we'll need to do another linter pass.

fujitatomoya commented on source/How-To-Guides/Using-ROSIndex.rst:
```diff
@@ -0,0 +1,51 @@
+Effectively Searching the ROS 2 Ecosystem
+#########################################
```
`Ecosystem` is kind ambiguous for me, what i can find here? CC: @kscottz WDYT?

```suggestion
Searching the ROS 2 Packages
############################
```

clalancette commented:
> I think this is a great beginner tutorial. We have a bit of a "librarian" problem where new users don't know about the resources available to them because they are spread out over half a dozen different websites. New ROS users do not understand that ROSIndex exists and that it is effectively PyPI for ROS packages. Given the current state of package-level docs on docs.ros.org we really need users using ROS Index to access their package-level documentation.

I'm in complete agreement that we should have *pointers* to the ROS index in prominent places.  I also agree that we haven't made it clear enough where the package documentation and the index are.

But I just am not convinced we need a tutorial explaining to people how to navigate the website.  We expect our users to be able to install Linux, use the command-line, and understand robotics concepts.  It just doesn't seem consistent to me to have to explain how a website works.

CursedRock17 committed:
t# This is a combination of 12 commits.

Update distribution spelling

Co-authored-by: Katherine Scott <katherineAScott@gmail.com>
Signed-off-by: Lucas Wendland <82680922+CursedRock17@users.noreply.github.com>

Update source/How-To-Guides/Using-ROSIndex.rst

Co-authored-by: Katherine Scott <katherineAScott@gmail.com>
Signed-off-by: Lucas Wendland <82680922+CursedRock17@users.noreply.github.com>

Update source/How-To-Guides/Using-ROSIndex.rst

Co-authored-by: Katherine Scott <katherineAScott@gmail.com>
Signed-off-by: Lucas Wendland <82680922+CursedRock17@users.noreply.github.com>

Update source/How-To-Guides/Using-ROSIndex.rst

Co-authored-by: Katherine Scott <katherineAScott@gmail.com>
Signed-off-by: Lucas Wendland <82680922+CursedRock17@users.noreply.github.com>

Update source/How-To-Guides/Using-ROSIndex.rst

Co-authored-by: Katherine Scott <katherineAScott@gmail.com>
Signed-off-by: Lucas Wendland <82680922+CursedRock17@users.noreply.github.com>

Update source/How-To-Guides/Using-ROSIndex.rst

Co-authored-by: Katherine Scott <katherineAScott@gmail.com>
Signed-off-by: Lucas Wendland <82680922+CursedRock17@users.noreply.github.com>

Update source/How-To-Guides/Using-ROSIndex.rst

Co-authored-by: Katherine Scott <katherineAScott@gmail.com>
Signed-off-by: Lucas Wendland <82680922+CursedRock17@users.noreply.github.com>

Update source/How-To-Guides/Using-ROSIndex.rst

Co-authored-by: Katherine Scott <katherineAScott@gmail.com>
Signed-off-by: Lucas Wendland <82680922+CursedRock17@users.noreply.github.com>

Update source/How-To-Guides/Using-ROSIndex.rst

Co-authored-by: Katherine Scott <katherineAScott@gmail.com>
Signed-off-by: Lucas Wendland <82680922+CursedRock17@users.noreply.github.com>

Update source/How-To-Guides/Using-ROSIndex.rst

Co-authored-by: Katherine Scott <katherineAScott@gmail.com>
Signed-off-by: Lucas Wendland <82680922+CursedRock17@users.noreply.github.com>

Update source/How-To-Guides/Using-ROSIndex.rst

Co-authored-by: Katherine Scott <katherineAScott@gmail.com>
Signed-off-by: Lucas Wendland <82680922+CursedRock17@users.noreply.github.com>

Update source/How-To-Guides/Using-ROSIndex.rst

Co-authored-by: Katherine Scott <katherineAScott@gmail.com>
Signed-off-by: Lucas Wendland <82680922+CursedRock17@users.noreply.github.com>

CursedRock17 committed:
Fixing underscores

Signed-off-by: CursedRock17 <mtglucas1@gmail.com>

Fixing hyperlink

Signed-off-by: CursedRock17 <mtglucas1@gmail.com>

Fixing odd link

Signed-off-by: CursedRock17 <mtglucas1@gmail.com>

Resolving list issue

Signed-off-by: CursedRock17 <mtglucas1@gmail.com>

CursedRock17 committed:
Resolve link

Signed-off-by: CursedRock17 <mtglucas1@gmail.com>

kscottz committed:
Update source/How-To-Guides/Using-ROSIndex.rst

Co-authored-by: Tomoya Fujita <Tomoya.Fujita@sony.com>
Signed-off-by: Katherine Scott <katherineAScott@gmail.com>

kscottz commented:
@clalancette I think you would be surprised. I wouldn't see this as, "here's how to use a website" as much as "here's the list of ROS packages and how you can find the information you need." 

We get a lot of questions that basically boil down to, "How do I find a package that does X?" It is helpful to have a page we point people to.

clalancette commented:
So we talked about this a little bit in one of the core maintainer meetings.  The consensus was that this kind of overview has value, but it probably belongs over at https://index.ros.org (maybe on the front page as a "New to index.ros.org?  Click here"), rather than on ROS 2 documentation.  Thinking about it more, that makes even more sense, since index.ros.org is a shared resource between ROS 1 and ROS 2.

So I'm going to suggest that what we do here is to close this PR, and instead open a PR that changes https://github.com/ros-infrastructure/rosindex/blob/ros2/index.html to add some of this information in instead.

Thoughts?

CursedRock17 commented:
Yeah I think that's fair, especially since [the main page](https://docs.ros.org/en/rolling/index.html#) of the docs, almost immedietaly says: 
> Are you looking for documentation for a particular ROS package like MoveIt, image_proc, or octomap? Please see [ROS Index](https://index.ros.org/packages/#rolling) or check out [this index of per-package documentation](https://docs.ros.org/en/rolling/p/).

I'm gonna go ahead and create a similar pull request in ROSIndex

marcoag commented:
Closing this PR as it's now moved to the rosindex repo.
